### DISCUSSION TITLE
webhook overloop POST

### DISCUSSION THREAD

**Discussion by newfaboc** [kirimwa-id/forum#66] [Tanya Jawab]
Hi rio,
saya sedang mencoba webook dan mendapat sedikit masalah.

saya mencoba post dengan method seperti ini : 

try {
  $reqParams = [
    'token' => getenv('API_TOKEN'),
    'url' => 'https://api.kirimwa.id/v1/webhooks',
    'method' => 'POST',
    'payload' => json_encode([
      'webhook_url' => 'https://mysite.com/my/webhook/coba.php'
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

di dalam coba.php tersebut saya memanggil kembali fungsi kirim send-message : 
try {
  $reqParams = [
    'token' => getenv('API_TOKEN'),
    'url' => 'https://api.kirimwa.id/v1/messages',
    'method' => 'POST',
    'payload' => json_encode([
      'message' => 'Halo ini adalah pesan dari api.kirimwa.id.',
      'phone_number' => '6281234567890',
      'message_type' => 'text',
      'device_id' => 'iphone-x-pro'
    ])
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

Namun dia mengirim pesan terus menerus.
Apakah ada yang salah ?

Saya juga mencoba menghapus webhook dengan metode : 
try {
  $reqParams = [
    'token' => getenv('API_TOKEN'),
    'url' => sprintf('https://api.kirimwa.id/v1/webhooks/%s', urlencode(getenv('WEBHOOK_ID'))),
    'method' => 'DELETE'
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

namun saya mendapati : 
Exception Object ( [message:protected] => {"message":"Error: Webhook not found."} [string:Exception:private] => [code:protected] => 404 [file:protected] ... 

saya sudah mencoba mengganti "getenv('WEBHOOK_ID')" dengan id yang saya dapatkan dari webhook yang saya coba POST namun tetap muncul pesan tersebut.

Mohon bantuannya.

Terima kasih

**Comment by rioastamal** [comment]
@newfaboc, Jika dalam `coba.php` anda mengirim pesan WA sedangkan URL `coba.php` dimasukkan dalam webhook maka akan terjadi **infinite loop** (loop yang tidak pernah berakhir).

Cek dulu list dari webhook dengan perintah `GET /v1/webhooks`. 

```json
{
  "data": [
    {
      "id": "whu-youremail@example.com",
      "data": "https://mywebhook.url/",
      "status": "active",
      "created_at": "2021-10-29T08:14:53.200Z"
    }
  ]
}
```

Pastikan yang diganti adalah `getenv("WEBHOOK_ID")` bukan hanya `WEBHOOK_ID` saja. Ganti dengan `whu-youremail@example.com`.

**Comment by newfaboc** [comment]
Ok. Saya akan coba dahulu.
Jika terjadi infinite loop, apakah ada cara untuk menghentikannya ?

**Reply by rioastamal** [reply]
Delete Webhook tersebut atau jangan mengirim pesan tanpa ada kondisi pada URL coba.php biar tidak infinite loop.

**Reply by newfaboc** [reply]
ok. sudah berhasil menghapus delete webhook.
Kemudian saya POST webhook ulang. 
Saat send-message WA, terkadang saya lihat GET webhook response tidak masuk. Apakah ada yang salah ?
metode yang saya gunakan :

try {
  $query = http_build_query([
    'status' => 'success',
    'webhook_type' => 'send_message_response',
    'start_key' => '',
    'device_id' => '',
    'message_id' => ''
  ]);
  $reqParams = [
    'token' => getenv('API_TOKEN'),
    'url' => sprintf('https://api.kirimwa.id/v1/webhook-responses?%s', $query)
  ];

  $response = apiKirimWaRequest($reqParams);
  echo $response['body'];
} catch (Exception $e) {
  print_r($e);
}

Apakah ada batasan log dari GET webhook-response ?
Bagaimana kah cara untuk mengclear log GET webhook-reponse ? 

Mohon bantuannya.

Terima kasih.

**Reply by rioastamal** [reply]
Mungkin statusnya `fail` bukan `success`. Coba ubah filternya.

Catatan: Webhook hanya dikirim ketika anda memiliki Webhook URL terdaftar jika anda menghapusnya maka webhook tidak terkirim.

**Reply by newfaboc** [reply]
oh. ternyata secara saya cek GET webhook-reponse untuk pengiriman Pesan dari API WA. sementara saya mencari incoming message.
Terima kasih mas rio :)

ada sedikit pertanyaan :
1. apakah ada perintah untuk mendapatkan LOG incoming message dengan webhook ?
2. apakah ada fungsi clear LOG untuk GET message, GET batch-message, GET webhook-response ?

**Reply by rioastamal** [reply]
1. https://developer.kirimwa.id/api/v1/#webhook-incoming-message ?
2. Tidak ada fungsi clear atau hapus log
